arch/arm/src/common/arm_interruptcontext.c

```diff
@@ -59,5 +59,15 @@
 
 bool up_interrupt_context(void)
 {
-  return CURRENT_REGS != NULL;
+#ifdef CONFIG_SMP
+  irqstate_t flags = up_irq_save();
```

Comment:
Just to clarify, it is non-atomic in the SMP case because CURRENT_REGS expands to:

    #  define CURRENT_REGS (g_current_regs[up_cpu_index()])

In the non-SMP case, it expands to:

    #  define CURRENT_REGS (g_current_regs[0])

Reply:
> Just to clarify, it is non-atomic in the SMP case because CURRENT_REGS expands to:
> 
> ```
> #  define CURRENT_REGS (g_current_regs[up_cpu_index()])
>```

@patacongo 

Thanks for the comments.

Yes, what I noticed at a Wi-Fi streaming aging test with Spresense was
NuttX stopped with DEBUGASSERT at the following code.

```
nxplayer> [ 4916.595630] cxd56_configure: ERROR: Unknown feature unit: 2048
[ 4916.598621] cxd56_configure: ERROR: Unknown feature unit: 4
[ 4916.604175] cxd56_configure: ERROR: Unknown feature unit: 16
[ 4982.154812] cxd56_start_dma: Underrun
[ 4982.205165] up_assert: Assertion failed CPU1 at file:semaphore/sem_wait.c line: 78 task: gs2200m

int nxsem_wait(FAR sem_t *sem)                                                                                                                                                                
{                                                                                                                                                                                             
  FAR struct tcb_s *rtcb = this_task();                                                                                                                                                       
  irqstate_t flags;                                                                                                                                                                           
  int ret = -EINVAL;                                                                                                                                                                          
                                                                                                                                                                                              
  /* This API should not be called from interrupt handlers */                                                                                                                                 
                                                                                                                                                                                              
=>  DEBUGASSERT(sem != NULL && up_interrupt_context() == false);    
````

Actually, the sem was not NULL but g_current_regs were the following values.

```
(gdb) p g_current_regs
$1 = {0x0, 0xd05df54}
```

Here, the DEBUGASSERT() happened on CPU1, and it was not in the interrupt context.
However, CPU0 was in the interrupt context, so I found that the task was moved from
CPU0 to CPU1 during calling up_interruput_context() in nxsem_wait().

With this PR, a Wi-Fi streaming aging test is still running over 12hrs.

Reply:
@patacongo 

This is the same as what I did for this_task() in sched/sched_thistask.c
So, may I merge this PR?

Reply:
> 
> 
> @patacongo
> 
> This is the same as what I did for this_task() in sched/sched_thistask.c
> So, may I merge this PR?

Well, we should never merge our own PRs.  But I suppose we should merge this now.  There is, however, a problem with the ARMv7-A with the GIC.  The use of up_irq_save() will not disable the SGIs used for inter-processor signalling, at least not until use of the ICCMPR is used as you suggested.

Perhaps we should add a comment about this in the PR?  What do you think?

Reply:
> There is, however, a problem with this change for the ARMv7-A with the GIC. The use of up_irq_save() will not disable the SGIs used for inter-processor signalling, at least not until use of the ICCMPR is used as you suggested. So I think that the fix does not work for that case.
> 
> Perhaps we should add a comment about this in the PR? What do you think?

@patacongo 
OK, I will add a comment.

Reply:
I already pushed the comment change and will merge in just a few minutes.  Thanks!  I did not expect you to still be working.  So I can handle the rest.